### PR TITLE
Log in JSON in addition to text files

### LINKED ISSUE #6: [Task 1]: Write the training logs to stdout.json as well so it can be parsed by HiPlot
**Task**: Currently, in VISSL, we training the training stats to the stdout and log.txt file however, these both are not directly parseable by tools like HiPlot. Further, having a json file allows anyone to interact with training and not require custom log.txt parsing code. Our goal in this task is to also log the `log_str` string here https://github.com/facebookresearch/vissl/blob/master/vissl/hooks/log_hooks.py#L210 to a `stdout.json` file. 

**Steps**:
1. Follow the example of metrics.json https://github.com/facebookresearch/vissl/blob/master/vissl/hooks/log_hooks.py#L402-L403 and create a `stdout.json` to which we will write `log_str` https://github.com/facebookresearch/vissl/blob/master/vissl/hooks/log_hooks.py#L210 instead 
2. Run CPU test `python tests/test_tasks.py` only. Monitor the logs closely and notice that `stdout.json` file is now saved as well.
3. Send a PR to VISSL and tag Priya as reviewer.

### PR DESCRIPTION
**Pull request by akainth015**
Resolves #6

### CODE REVIEW

File: tests/test_tasks.py [hunk 1]
```diff
@@ -5,13 +5,14 @@
...
 import pkg_resources
 from parameterized import parameterized
-from utils import UNIT_TEST_CONFIGS, SSLHydraConfig
 from vissl.engines.train import train_main
 from vissl.hooks import default_hook_generator
 from vissl.utils.checkpoint import get_checkpoint_folder
 from vissl.utils.hydra_config import convert_to_attrdict
 from vissl.utils.misc import get_dist_run_id
 
+from .utils import UNIT_TEST_CONFIGS, SSLHydraConfig
```
Comment (prigoyal) [inline tests/test_tasks.py]: just curious: what's the motivation for this change? does anything break if we don't specify the "." aka the relative import?

File: .gitignore [hunk 2]
```diff
@@ -35,6 +35,7 @@ var/
 .hydra/
 tb_logs/
 metrics.json
+stdout.json
```
Comment (prigoyal) [inline .gitignore:38]: nice!

File: vissl/hooks/log_hooks.py [hunk 3]
```diff
@@ -191,6 +191,16 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
                     f"eta: {eta_string}; "
                     f"peak_mem: {peak_mem_used}M"
                 )
+                log_data = {
+                    "Rank": rank,
+                    "ep": train_phase_idx,
+                    "iter": iteration,
+                    "lr": lr_val,
+                    "loss": loss_val,
+                    "btime": batch_time,
```
Comment (prigoyal) [inline vissl/hooks/log_hooks.py]: nit: `btime(ms)`

File: vissl/hooks/log_hooks.py [hunk 4]
```diff
@@ -191,6 +191,16 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
...
+                    "eta": eta_string,
+                    "peak_mem": peak_mem_used,
```
Comment (prigoyal) [inline vissl/hooks/log_hooks.py]: nit: `peak_mem(M)`

File: vissl/hooks/log_hooks.py [hunk 5]
```diff
@@ -207,7 +217,10 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
                         f"btime({self.btime_freq}iters): {rolling_btime} ms; "
                         f"rolling_eta: {rolling_eta_str}"
                     )
+                    log_data["btime_iters"] = rolling_btime
+                    log_data["rolling_eta"] = rolling_eta_str
                 logging.info(log_str)
+                save_file(log_data, f"{task.checkpoint_folder}/stdout.json")
```
Comment (prigoyal) [inline vissl/hooks/log_hooks.py]: I just realized that the call to `save_file` will consistently spew the logs https://github.com/MLH-Fellowship/vissl/blob/master/vissl/utils/io.py#L56 which will be noisy. Yet we can't disable the logging in L56 in `save_file`. I recommend to replace the `save_file(...)` with 
```
with PathManager.open(f"{task.checkpoint_folder}/stdout.json", "a") as fopen:
        fopen.write(json.dumps(data, sort_keys=False) + "\n")
        fopen.flush()
```

File: vissl/hooks/log_hooks.py [hunk 6]
```diff
@@ -191,6 +191,16 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
                     f"eta: {eta_string}; "
```
Comment (prigoyal) [inline vissl/hooks/log_hooks.py]: let's remove the `log_str` from lines 184 to line 193 in favor of `log_data`

File: vissl/hooks/log_hooks.py [hunk 7]
```diff
@@ -207,7 +217,10 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
                         f"btime({self.btime_freq}iters): {rolling_btime} ms; "
                         f"rolling_eta: {rolling_eta_str}"
                     )
+                    log_data["btime_iters"] = rolling_btime
```
Comment (prigoyal) [inline vissl/hooks/log_hooks.py]: nit: `log_data[f"btime({self.btime_freq}iters) (ms)"] = rolling_btime`

File: vissl/hooks/log_hooks.py [hunk 8]
```diff
@@ -207,7 +217,10 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
...
+                    log_data["rolling_eta"] = rolling_eta_str
                 logging.info(log_str)
```
Comment (prigoyal) [inline vissl/hooks/log_hooks.py]: let's replace this line with `logging.info(json.dumps(log_data))` and completely get rid of `log_str`

**Review by prigoyal — CHANGES_REQUESTED** [review]
thank you for the PR @akainth015 and @grace-omotoso , looks like a great start.

I have few comments inline that needs to be addressed.

Additionally, can you update the PR with the test plan: show me the output of the `stdout.json` file that gets created.

File: tests/test_tasks.py (hunk 1 above)
Comment (akainth015) [reply tests/test_tasks.py]: Yes, I was planning on bringing it up during the meeting. The test command you gave me (`python3 tests/test_tasks.py`) did not run anything, so I used `python3 -m unittest test/test_tasks.py` instead. There was an `ImportError` while running it without `.utils`, so I added it to both of our pull requests.

**Comment by akainth015** [comment]
@prigoyal I noticed that repeated runs will append to the existing `metrics.json` and `stdout.json` without any explicit indication that a whole new execution being logged. For example, running the tests yields 4 lines in `metrics.json`, running again will give 8, 12, and so on.

**Comment by prigoyal** [comment]
> @prigoyal I noticed that repeated runs will append to the existing `metrics.json` and `stdout.json` without any explicit indication that a whole new execution being logged. For example, running the tests yields 4 lines in `metrics.json`, running again will give 8, 12, and so on.

yep, that's expected when your output directory is the same across repeated runs. You can clear out the checkpoint directory for now !

File: vissl/hooks/log_hooks.py (hunk 3 above)
Comment (akainth015) [reply vissl/hooks/log_hooks.py]: The reason I initially renamed the keys is that if these logs are consumed by JavaScript scripts, etc. they would have to use the `log["btime(ms)"]` format instead of the simpler `log.btime` dot notation. Do you think including the units in the key (since that would be repeated in every logged frame) is better than sticking the units onto some documentation page?

**Comment by akainth015** [comment]
`stdout.json` from `python3 -m unittest tests/test_tasks.py`
```
{"Rank": 0, "ep": 0, "iter": 1, "lr": 0.00234, "loss": 3.00722, "btime(ms)": 765, "eta": "0:00:00", "peak_mem(M)": -1}
{"Rank": 0, "ep": 0, "iter": 1, "lr": 0.00047, "loss": 6.38364, "btime(ms)": 660, "eta": "0:00:01", "peak_mem(M)": -1}
{"Rank": 0, "ep": 0, "iter": 1, "lr": 0.00234, "loss": 1.3317, "btime(ms)": 1199, "eta": "0:00:01", "peak_mem(M)": -1}
{"Rank": 0, "ep": 0, "iter": 1, "lr": 0.00234, "loss": 1.13592, "btime(ms)": 382, "eta": "0:00:00", "peak_mem(M)": -1}
```

File: vissl/hooks/log_hooks.py (hunk 3 above)
Comment (prigoyal) [reply vissl/hooks/log_hooks.py]: that's a very good insight @akainth015 :)

We would want to stick with the units in the key since these are quite critical aspects to how we monitor the trainings and units can often become very confusing. Hence we intentionally listed the units from the very beginning.

I see your point though. If we encounter a need in future for Javascript etc, we will revisit this :)

File: vissl/data/ssl_transforms/rand_auto_aug.py [hunk 9]
```diff
@@ -69,7 +69,7 @@
 
 
 # Modification/Addition
-@register_transform("VisslRandAugment")
+@register_transform("RandAugment")
```
Comment (prigoyal) [inline vissl/data/ssl_transforms/rand_auto_aug.py]: we should rebase to master. This change should be essentially reverted after the rebase.

File: vissl/config/defaults.yaml [hunk 10]
```diff
@@ -195,7 +195,7 @@ config:
       # no valid values other than 0 or None), unclear if beneficial. Default =
       # None.
       # TRANSFORMS:
-      #   - name: VisslRandAugment
+      #   - name: RandAugment
```
Comment (prigoyal) [inline vissl/config/defaults.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/debugging/pretrain/supervised/supervised_1gpu_vision_transformer_debug_integration_cutmix.yaml [hunk 11]
```diff
@@ -19,7 +19,7 @@ config:
       TRANSFORMS:
         - name: RandomResizedCrop
           size: 224
-        - name: VisslRandAugment
+        - name: RandAugment
```
Comment (prigoyal) [inline configs/config/debugging/pretrain/supervised/supervised_1gpu_vision_transformer_debug_integration_cutmix.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/debugging/pretrain/supervised/supervised_1gpu_vision_transformer_debug_integration_cutmix.yaml [hunk 12]
```diff
@@ -35,7 +35,6 @@ config:
       }
     TEST:
       DATA_SOURCES: [disk_folder]
-      #      DATA_PATHS: ["<path to test folder>"]
```
Comment (prigoyal) [inline configs/config/debugging/pretrain/supervised/supervised_1gpu_vision_transformer_debug_integration_cutmix.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/debugging/pretrain/supervised/supervised_deit_b_integration_debug.yaml [hunk 13]
```diff
@@ -20,7 +20,7 @@ config:
         - name: RandomResizedCrop
           size: 224
         - name: RandomHorizontalFlip
-        - name: VisslRandAugment
+        - name: RandAugment
```
Comment (prigoyal) [inline configs/config/debugging/pretrain/supervised/supervised_deit_b_integration_debug.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/pretrain/vision_transformer/supervised/supervised_16gpu_deit_b_example.yaml [hunk 14]
```diff
@@ -16,11 +16,12 @@ config:
       LABEL_SOURCES: [disk_folder]
       DATASET_NAMES: [imagenet1k_folder]
       BATCHSIZE_PER_REPLICA: 32
+      DROP_LAST: True
```
Comment (prigoyal) [inline configs/config/pretrain/vision_transformer/supervised/supervised_16gpu_deit_b_example.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/pretrain/vision_transformer/supervised/supervised_16gpu_deit_b_example.yaml [hunk 15]
```diff
@@ -16,11 +16,12 @@ config:
...
       TRANSFORMS:
         - name: RandomResizedCrop
           size: 224
         - name: RandomHorizontalFlip
-        - name: VisslRandAugment
+        - name: RandAugment
```
Comment (prigoyal) [inline configs/config/pretrain/vision_transformer/supervised/supervised_16gpu_deit_b_example.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml [hunk 16]
```diff
@@ -3,7 +3,6 @@ config:
   HOOKS:
     TENSORBOARD_SETUP:
       USE_TENSORBOARD: True
-      EXPERIMENT_LOG_DIR: "/checkpoint/ito/vision_transformer/1gpu_test"
```
Comment (prigoyal) [inline configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml [hunk 17]
```diff
@@ -42,14 +40,15 @@ config:
...
+        "ssl_method": "swav",
         "mixup_alpha": 1.0, # mixup alpha value, mixup is active if > 0.
         "cutmix_alpha": 1.0, # cutmix alpha value, cutmix is active if > 0.
         "prob": 1.0, # probability of applying mixup or cutmix per batch or element
         "switch_prob": 0.5, # probability of switching to cutmix instead of mixup when both are active
         "mode": "batch", # how to apply mixup/cutmix params (per 'batch', 'pair' (pair of elements), 'elem' (element)
         "correct_lam": True, # apply lambda correction when cutmix bbox clipped by image borders
         "label_smoothing": 0.1, # apply label smoothing to the mixed target tensor
-        "num_classes": 1000 # number of classes for target
+        "num_classes": 1 # number of classes for target
```
Comment (prigoyal) [inline configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml [hunk 18]
```diff
@@ -13,19 +12,18 @@ config:
...
+      DROP_LAST: True
       TRANSFORMS:
         - name: ImgPilToMultiCrop
           total_num_crops: 2
           size_crops: [224]
           num_crops: [2]
           crop_scales: [[0.14, 1]]
         - name: RandomHorizontalFlip
-        - name: VisslRandAugment
+        - name: RandAugment
```
Comment (prigoyal) [inline configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

File: configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml [hunk 19]
```diff
@@ -13,19 +12,18 @@ config:
     NUM_DATALOADER_WORKERS: 8
     TRAIN:
       DATA_SOURCES: [disk_folder]
-      #      DATA_PATHS: ["<path to train folder>"]
-      LABEL_SOURCES: [disk_folder]
-      DATASET_NAMES: [imagenet1k_debug_folder]
+      DATASET_NAMES: [imagenet1k_folder]
```
Comment (prigoyal) [inline configs/config/pretrain/vision_transformer/swav/swav_deit_b_2nodes.yaml]: we should rebase to master. This change should be essentially reverted after the rebase.

**Review by prigoyal — COMMENTED** [review]
Looks good to go to me!

There are a couple of changes that should be removed after we rebase to current master. 

Thank you @akainth015 , looking very good :)

**Review by uhbuhb — COMMENTED** [review]
Good job. A couple of general things:

This PR can be split into 3 topics:
1. changing log_str to log_data and logging it to a json file instead of stdout.
2. updates to unittests
3. .gitignore the log.json file.

Ideally the commits in the PR should reflect these changes clearly (so that it will be possible to revert specific commits if necessary, but also for readability). 

Regarding the middle change of the unittests... sometimes changing absolute to relative imports can cause unexpected problems.. may be worthwhile getting to the root reason why they werent working as is (or how they do work as is for someone else) and then considering if in both cases they will continue to work after changing to relative imports. Could be because of the directory the tests are run from, which affects the pythonpath and therefore the import statements..

Other than that - rebase and great job!

**Comment by prigoyal** [comment]
1. Rebase
2. commit split
3. after steps 1, 2 run a test (remove the checkpoints folder, existing json) and paste the test command + output of stdout.json in the PR open box. 
4. Ensure CircleCI tests are green
5. tag Priya to merge the PR

**Comment by prigoyal** [comment]
this PR still requires rebase :) it seems to be 5 commits behind the upstream/master :)

File: tests/test_tasks.py [hunk 20]
```diff
@@ -5,13 +5,14 @@
...
 import pkg_resources
 from parameterized import parameterized
-from utils import UNIT_TEST_CONFIGS, SSLHydraConfig
 from vissl.engines.train import train_main
 from vissl.hooks import default_hook_generator
 from vissl.utils.checkpoint import get_checkpoint_folder
 from vissl.utils.hydra_config import convert_to_attrdict
 from vissl.utils.misc import get_dist_run_id
 
+from utils import UNIT_TEST_CONFIGS, SSLHydraConfig
```
Comment (prigoyal) [inline tests/test_tasks.py]: one small comment: this file should be untouched in the commit history since we don't need the relative import. Are the changes stemming from the isort ? otherwise let's revert any changes to this file and good to press merge right away then! :)

**Review by prigoyal — APPROVED** [review]
looks great to me! thank you so much :) merging it

**Comment by prigoyal** [comment]
@akainth015 , @grace-omotoso , sharing the the test to run for ensuring the "epoch" is correctly changed as well 
`python tools/run_distributed_engines.py config=test/cpu_test/test_cpu_resnet_simclr config.DATA.TRAIN.DATA_LIMIT=50`
